conda/common/iterators.py

```diff
@@ -18,3 +21,24 @@ def groupby_to_dict(keyfunc, sequence):
     for key, group in itertools.groupby(sequence, keyfunc):
         result[key].extend(group)
     return dict(result)
+
+
+T = TypeVar("T")
+
+
+def next_instance(objects: Iterable[T], object_type: type[T]) -> T | None:
```

Comment:
The typing is too strict, there is no requirement that the iterable elements must be an instance of `T`:

```suggestion
def next_instance(objects: Iterable[object], object_type: type[T] | tuple[type[T], ...]) -> T | None:
```

We could also reuse what [typeshed](https://github.com/python/typeshed/blob/7c15e6998f28272a5e308524dadfc5aa00aff08b/stdlib/builtins.pyi#L1327-L1332) does?

What do you think about replacing `object_type` with `classinfo` to mirror `isinstance`/`issubclass`?

Reply:
Good catch. I think I will just re-write it as:

```python
from typing import Any

# snip, snip...

def next_instance(objects: Iterable[Any], object_type: type[T]) -> T | None:
```

That way it will be more generic.

I wouldn't want the second argument to accept multiple values though (i.e. `tuple[type[T], ...]`). The function is supposed to return the _next_ instance of `type`, which means just one single instance. Being able to define multiple types just doesn't make much sense to me. Would it just return the first type found in the tuple? If that's the case, why not just pass in a single type to begin with. Hope that makes sense.